File: harness/oracle/cases/room-characters.ts
/**
 * Oracle case (P4.D172): the room's character map — one batched read, every
 * present seat (v4 `d14da3a56`, bug 131).
 *
 * Drives v4's REAL `lib/chat/turn-manager/room-characters.ts` at the
 * `78b381a96` pin over a counting stand-in for `repos.characters.findByIds`,
 * exactly as v4's own `room-characters.test.ts` does. The stub is what makes the
 * CALL COUNT a comparand: "reads the whole room in ONE call, not one per seat"
 * is the point of the commit, and only the count can say it.
 *
 * The bug this closes: four of the six turn paths built their map from
 * `getActiveCharacterParticipants`, which returns LLM seats only, so a seat the
 * human drives never reached the map — its talkativeness fell through to 0.5 and
 * its `archivedAt` was invisible to the seat filter.
 *
 * v4's own talkativeness case is STATISTICAL (300 loud vs 300 quiet). It cannot
 * be a differential row; the `bug131-*` draw rows below make the same point
 * deterministically with pinned draws, and the Rust side mirrors v4's bounds.
 *
 * Concrete-module imports, not the barrel — see `cycle-order.ts`'s note.
 *
 * Run from inside the server checkout:
 *   cd ~/source/quilltap-server
 *   npx tsx ~/source/quilltap-v5/harness/oracle/cases/room-characters.ts \
 *     > /tmp/oracle-room-characters.ndjson
 */

import { loadRoomCharacters } from '@/lib/chat/turn-manager/room-characters';
import { cycleCandidates, drawCycleOrder } from '@/lib/chat/turn-manager/cycle-order';
import type { ChatParticipantBase, Character } from '@/lib/schemas/types';

const NOW = '2026-09-09T00:00:00.000Z';

type WirePart = {
  id: string;
  type: string;
  status: string;
  characterId: string | null;
  controlledBy: string;
  talkativeness: number | null;
};
type WireChar = { id: string; name?: string; talkativeness?: number | null; archivedAt?: string | null };

const p = (
  id: string,
  characterId: string | null,
  controlledBy = 'llm',
  status = 'active',
  type = 'CHARACTER',
  talkativeness: number | null = null,
): WirePart => ({ id, type, status, characterId, controlledBy, talkativeness });

const asParts = (ps: WirePart[]) => ps as unknown as ChatParticipantBase[];
const asChar = (c: WireChar) =>
  ({ createdAt: NOW, updatedAt: NOW, name: c.name ?? 'Character ' + c.id, ...c } as unknown as Character);

/** The counting stand-in for the batched read — v4's own `makeRepos`. */
const makeRepos = (rows: WireChar[]) => {
  const calls: string[][] = [];
  const repos = {
    characters: {
      findByIds: async (ids: string[]) => {
        calls.push([...ids]);
        return rows.filter((c) => ids.includes(c.id)).map(asChar);
      },
    },
  };
  return { repos, calls };
};

type LoadCase = {
  id: string;
  participants: WirePart[];
  rows: WireChar[];
  preloaded?: Array<WireChar | null>;
};

const loadCases: LoadCase[] = [
  {
    id: 'includes-user-driven-seats-not-just-llm',
    participants: [p('p-llm', 'char-llm'), p('p-user', 'char-user', 'user')],
    rows: [{ id: 'char-llm' }, { id: 'char-user', talkativeness: 0.9 }],
  },
  {
    id: 'reads-the-whole-room-in-one-call',
    participants: [p('p1', 'char-1'), p('p2', 'char-2'), p('p3', 'char-3'), p('p4', 'char-4', 'user')],
    rows: [{ id: 'char-1' }, { id: 'char-2' }, { id: 'char-3' }, { id: 'char-4' }],
  },
  {
    id: 'asks-only-for-present-seats-carrying-a-character',
    participants: [
      p('p-active', 'char-1'),
      p('p-silent', 'char-2', 'llm', 'silent'),
      p('p-left', 'char-3', 'llm', 'absent'),
      p('p-removed', 'char-5', 'llm', 'removed'),
      p('p-nochar', null),
      p('p-narrator', 'char-4', 'llm', 'active', 'NARRATOR'),
    ],
    rows: [{ id: 'char-1' }, { id: 'char-2' }, { id: 'char-3' }, { id: 'char-4' }, { id: 'char-5' }],
  },
  {
    id: 'dedupes-two-seats-playing-the-same-character',
    participants: [p('p1', 'char-1'), p('p2', 'char-1')],
    rows: [{ id: 'char-1' }],
  },
  {
    id: 'skips-the-read-entirely-when-no-seat-is-present',
    participants: [p('p-left', 'char-1', 'llm', 'absent')],
    rows: [{ id: 'char-1' }],
  },
  {
    id: 'leaves-an-unreadable-character-out-of-the-map',
    participants: [p('p1', 'char-1'), p('p2', 'char-shelved')],
    rows: [{ id: 'char-1' }],
  },
  {
    id: 'prefers-a-preloaded-character-over-the-copy-the-read-returned',
    participants: [p('p1', 'char-1')],
    rows: [{ id: 'char-1', name: 'Stale' }],
    preloaded: [{ id: 'char-1', name: 'Fresh' }],
  },
  {
    id: 'ignores-empty-preloaded-entries',
    participants: [p('p1', 'char-1')],
    rows: [{ id: 'char-1', name: 'Real' }],
    preloaded: [null, null],
  },
  {
    // A preloaded character for a seat the room does NOT hold is still seeded
    // (v4 gates only on `preloaded?.id`), which is why the finalizer's
    // just-spoke seeding is safe even mid-cast-change.
    id: 'a-preloaded-stranger-is-still-seeded',
    participants: [p('p1', 'char-1')],
    rows: [{ id: 'char-1' }],
    preloaded: [{ id: 'char-stranger', name: 'Stranger' }],
  },
  {
    id: 'an-empty-room-reads-nothing',
    participants: [],
    rows: [{ id: 'char-1' }],
  },
];

const rows: unknown[] = [];

for (const c of loadCases) {
  const { repos, calls } = makeRepos(c.rows);
  const map = await loadRoomCharacters(
    repos,
    asParts(c.participants),
    c.preloaded ? { preloaded: c.preloaded.map((x) => (x ? asChar(x) : null)) } : undefined,
  );
  rows.push({
    kind: 'load',
    id: c.id,
    participants: c.participants,
    rows: c.rows,
    preloaded: c.preloaded ?? null,
    // The map, projected to what every consumer reads plus the name the chain
    // decision reads. Key ORDER is not compared (both sides are hash maps); the
    // key SET and each value are.
    out: Object.fromEntries(
      [...map.entries()].map(([id, ch]) => [
        id,
        { name: ch.name, talkativeness: ch.talkativeness ?? null, archivedAt: ch.archivedAt ?? null },
      ]),
    ),
    // The point of the commit: how many times the batched read was called, and
    // with which ids (in v4's insertion order).
    calls,
  });
}

// The two consumption arms — bug 131's actual symptoms, over a map that can see
// the user seat.
type ConsumeCase = { id: string; participants: WirePart[]; rows: WireChar[] };
const consumeCases: ConsumeCase[] = [
  {
    id: 'keeps-a-seat-whose-character-could-not-be-read',
    participants: [p('p1', 'char-1'), p('p2', 'char-shelved')],
    rows: [{ id: 'char-1' }],
  },
  {
    id: 'excludes-a-user-seat-whose-character-is-archived',
    participants: [p('p-llm', 'char-llm'), p('p-user', 'char-user', 'user')],
    rows: [{ id: 'char-llm' }, { id: 'char-user', archivedAt: NOW }],
  },
];
for (const c of consumeCases) {
  const { repos } = makeRepos(c.rows);
  const map = await loadRoomCharacters(repos, asParts(c.participants));
  rows.push({
    kind: 'candidates',
    id: c.id,
    participants: c.participants,
    rows: c.rows,
    out: cycleCandidates(asParts(c.participants), map).map((x) => x.id),
  });
}

// bug 131 deterministically: the SAME room drawn with the SAME pinned draw, once
// with a loud user seat and once with a quiet one. A map built from LLM seats
// alone weights the human at 0.5 either way, so the two rows come out identical;
// only a whole-room map separates them. (v4's own case measures this over 300
// draws; these two rows are the same claim with the RNG held still.)
for (const [id, userTalk, pin] of [
  // The pin is CHOSEN, not arbitrary: at 0.8 the whole-room map gives
  // `p-user` for the loud seat and `p-llm-2` for the quiet one, while an
  // LLM-only map — which weights the human at the 0.5 default whatever their
  // character says — gives `p-user` for BOTH. So the two rows differing IS the
  // fix, and their agreeing is the bug.
  ['bug131-loud-user-seat', 0.98, 0.8],
  ['bug131-quiet-user-seat', 0.02, 0.8],
] as Array<[string, number, number]>) {
  const participants = [
    p('p-llm-1', 'char-llm-1'),
    p('p-llm-2', 'char-llm-2'),
    p('p-user', 'char-user', 'user'),
  ];
  const rowsIn: WireChar[] = [
    { id: 'char-llm-1', talkativeness: 0.98 },
    { id: 'char-llm-2', talkativeness: 0.98 },
    { id: 'char-user', talkativeness: userTalk },
  ];
  const { repos } = makeRepos(rowsIn);
  const map = await loadRoomCharacters(repos, asParts(participants));
  const orig = Math.random;
  const consumed: number[] = [];
  Math.random = () => {
    consumed.push(pin);
    return pin;
  };
  let order: string[];
  try {
    order = drawCycleOrder({ participants: asParts(participants), characters: map });
  } finally {
    Math.random = orig;
  }
  rows.push({ kind: 'draw', id, participants, rows: rowsIn, draws: [pin], out: order, consumedDraws: consumed });
}

for (const r of rows) process.stdout.write(JSON.stringify(r) + '\n');

File: crates/quilltap-core/src/room_characters.rs
//! Port of v4's lib/chat/turn-manager/room-characters.ts (`d14da3a56`, bug 131)
//! — the one place a turn path loads "who is in this room".
//!
//! Every server path that asks "who speaks next" needs a `characterId →
//! Character` map first: [`crate::cycle_order::resolve_cycle_order_pure`] weights
//! the rotation draw by talkativeness out of it, and
//! [`crate::select_speaker::select_next_speaker`] reads `archivedAt` out of it to
//! drop tombstoned seats. Six call sites used to build that map by hand, and four
//! of them built it from `getActiveCharacterParticipants` — which, despite the
//! name, returns only LLM-controlled seats. A seat the human drives was therefore
//! invisible to both readers: its character's talkativeness never reached the draw
//! (it fell through to the 0.5 default unless the seat carried a per-chat
//! override), and an archived character on it was never excluded. Building the map
//! over [`crate::participant_filters::get_present_character_seats`] instead is what
//! this module is for.
//!
//! It is also the cheaper read, which is why widening the map costs nothing.
//! `find_by_id` overlays one character's vault through `loadVaultFileMaps([one
//! mount])` — eleven vault queries plus the row — so the old per-seat loops paid
//! that eleven times over in a four-seat room. `find_by_ids` overlays the whole
//! room in one pass, whatever the seat count.
//!
//! **Best-effort, by contract.** [`crate::cycle_order::cycle_candidates`]
//! documents its input as a best-effort read and deliberately KEEPS a seat whose
//! character is missing from the map, so that a failed lookup costs a name rather
//! than silently shrinking the room. The batched read is what actually honours
//! that: the list overlay logs and DROPS a character whose vault is unavailable,
//! where the single-character overlay behind `find_by_id` returns an error —
//! which, in the old loops, propagated straight out of speaker selection and took
//! the whole turn (and the read-only `?action=turn` projection behind the
//! participant sidebar) with it.

use std::collections::HashMap;

use serde_json::Value;

use crate::db::DbError;
use crate::participant_filters::is_present_character_seat;
use crate::select_speaker::{SpeakerCharacter, SpeakerParticipant};

/// The batched read [`load_room_characters`] needs — v4's
/// `repos.characters.findByIds`.
///
/// A closure rather than a trait so the differential can count calls on the Rust
/// side exactly as v4's `jest.fn` counts them on its own, and so the production
/// caller can pass its nested `read_main(read_mount_index(…))` without a wrapper
/// type.
pub type FindCharactersByIds<'a> = &'a mut dyn FnMut(&[String]) -> Result<Vec<Value>, DbError>;

/// Loads `characterId → Character` for every character seat present in the room —
/// LLM-driven and user-driven alike — in a single batched read.
///
/// Seats whose character cannot be read are simply absent from the map. That is
/// the documented contract of every consumer: an unknown character is treated as
/// present and unarchived at the default weight, never as a reason to drop a seat
/// or fail a turn.
///
/// `preloaded` characters are seeded AFTER the batch, so the caller's copy WINS.
/// The message finalizer uses this for the character that just spoke: it has the
/// authoritative post-turn record in hand and must not be handed a staler one.
/// Entries without a (truthy) `id` are ignored, matching v4's `if (preloaded?.id)`.
pub fn load_room_characters(
    participants: &[SpeakerParticipant],
    preloaded: &[Value],
    find_by_ids: FindCharactersByIds<'_>,
) -> Result<HashMap<String, Value>, DbError> {
    let seats: Vec<&SpeakerParticipant> = participants
        .iter()
        .filter(|p| {
            is_present_character_seat(&p.participant_type, p.status, p.character_id.as_deref())
        })
        .collect();

    // v4 `Array.from(new Set(seats.map(characterId)))` — deduped, in first-seen
    // order (two seats can play the same character).
    let mut ids: Vec<String> = Vec::new();
    for s in &seats {
        let cid = s.character_id.clone().unwrap_or_default();
        if !ids.contains(&cid) {
            ids.push(cid);
        }
    }

    let mut characters: HashMap<String, Value> = HashMap::new();

    // v4 skips the read entirely for an empty room — the call COUNT is a
    // comparand, so this guard is behaviour, not an optimisation.
    if !ids.is_empty() {
        for row in find_by_ids(&ids)? {
            if let Some(id) = row.get("id").and_then(Value::as_str) {
                characters.insert(id.to_string(), row);
            }
        }
    }

    for p in preloaded {
        if let Some(id) = p
            .get("id")
            .and_then(Value::as_str)
            .filter(|s| !s.is_empty())
        {
            characters.insert(id.to_string(), p.clone());
        }
    }

    tracing::debug!(
        seats = seats.len(),
        requested = ids.len(),
        resolved = characters.len(),
        "[Turn Manager] Room characters loaded"
    );

    // A shortfall means a seat's character row is gone or its vault is on the
    // shelf. Worth saying out loud — the room is about to be weighted and ordered
    // without it — but never worth failing a turn over.
    let missing: Vec<&String> = ids
        .iter()
        .filter(|id| !characters.contains_key(*id))
        .collect();
    if !missing.is_empty() {
        tracing::warn!(
            requested = ids.len(),
            missing = ?missing,
            "[Turn Manager] Room seats with no readable character"
        );
    }

    Ok(characters)
}

/// Project the room map down to what the selection and the rotation draw read —
/// v4's `Map<string, Character>` narrowed to `talkativeness` + `archivedAt`.
///
/// Kept beside the loader so the two facts always travel together: a parallel
/// "archived ids" argument could disagree with the map, and a caller that forgot
/// it would silently let a tombstone take turns
/// ([`crate::select_speaker::SpeakerCharacter`]'s own note).
pub fn to_speaker_characters(room: &HashMap<String, Value>) -> HashMap<String, SpeakerCharacter> {
    room.iter()
        .map(|(id, row)| {
            (
                id.clone(),
                SpeakerCharacter {
                    talkativeness: row.get("talkativeness").and_then(Value::as_f64),
                    archived: crate::api::characters::is_archived(row),
                },
            )
        })
        .collect()
}

/// The `name` of a room character, for the chain decision's `characterName`
/// (v4 `charactersMap.get(id)?.name`).
pub fn room_character_name<'a>(
    room: &'a HashMap<String, Value>,
    character_id: Option<&str>,
) -> Option<&'a str> {
    character_id
        .filter(|c| !c.is_empty())
        .and_then(|cid| room.get(cid))
        .and_then(|c| c.get("name"))
        .and_then(Value::as_str)
}

/// [`load_room_characters`] over the real repositories — the shape every server
/// selection site uses (v4's `loadRoomCharacters(repos, participants, …)`).
///
/// The nested `read_main(read_mount_index(…))` is what
/// `characters_read::find_by_ids` needs for its vault overlay; keeping it here
/// means the six call sites do not each re-derive it.
pub fn load_room_characters_from_db(
    db: &crate::db::runtime::Db,
    participants: &[SpeakerParticipant],
    preloaded: &[Value],
) -> Result<HashMap<String, Value>, DbError> {
    let mut find_by_ids = |ids: &[String]| -> Result<Vec<Value>, DbError> {
        db.read_main(|main| {
            db.read_mount_index(|mount| crate::db::characters_read::find_by_ids(main, mount, ids))
        })
    };
    load_room_characters(participants, preloaded, &mut find_by_ids)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::chat_predicates::ParticipantStatus;

    fn seat(id: &str, character_id: Option<&str>, controlled_by: &str) -> SpeakerParticipant {
        SpeakerParticipant {
            id: id.to_string(),
            participant_type: "CHARACTER".to_string(),
            status: ParticipantStatus::Active,
            character_id: character_id.map(String::from),
            controlled_by: controlled_by.to_string(),
            talkativeness: None,
        }
    }

    fn row(id: &str) -> Value {
        serde_json::json!({ "id": id, "name": format!("Character {id}") })
    }

    // v4 `room-characters.ts:85-89`, on the hot turn path.
    #[test]
    fn the_loaded_line_carries_v4s_three_counts() {
        let parts = vec![
            seat("p1", Some("c1"), "llm"),
            seat("p2", Some("c1"), "user"),
        ];
        let lines = crate::test_support::captured(|| {
            let mut find = |_ids: &[String]| Ok(vec![row("c1")]);
            let map = load_room_characters(&parts, &[], &mut find).unwrap();
            assert_eq!(map.len(), 1);
        });
        // Two seats, ONE requested id (deduped), one resolved.
        let hit = lines
            .iter()
            .find(|l| l.contains("[Turn Manager] Room characters loaded"))
            .unwrap_or_else(|| panic!("no loaded line in {lines:?}"));
        assert!(hit.contains("seats=2"), "{hit}");
        assert!(hit.contains("requested=1"), "{hit}");
        assert!(hit.contains("resolved=1"), "{hit}");
    }

    // v4 `room-characters.ts:96-99` — and its SILENCE arm, without which the
    // assertion above passes on a logger that shouts about everything.
    #[test]
    fn the_shortfall_warns_and_a_full_room_does_not() {
        let parts = vec![
            seat("p1", Some("c1"), "llm"),
            seat("p2", Some("shelved"), "llm"),
        ];
        let short = crate::test_support::captured(|| {
            let mut find = |_ids: &[String]| Ok(vec![row("c1")]);
            load_room_characters(&parts, &[], &mut find).unwrap();
        });
        let hit = short
            .iter()
            .find(|l| l.contains("[Turn Manager] Room seats with no readable character"))
            .unwrap_or_else(|| panic!("no shortfall warn in {short:?}"));
        assert!(hit.contains("requested=2"), "{hit}");
        assert!(hit.contains("shelved"), "{hit}");

        let full = crate::test_support::captured(|| {
            let mut find = |_ids: &[String]| Ok(vec![row("c1"), row("shelved")]);
            load_room_characters(&parts, &[], &mut find).unwrap();
        });
        assert!(
            !full
                .iter()
                .any(|l| l.contains("Room seats with no readable character")),
            "a fully-resolved room must not warn: {full:?}"
        );
    }

    // The read error is NOT swallowed: v4's `await` propagates, and so must this.
    #[test]
    fn a_failed_read_propagates() {
        let parts = vec![seat("p1", Some("c1"), "llm")];
        let mut find = |_ids: &[String]| Err(DbError::Internal("boom".into()));
        assert!(load_room_characters(&parts, &[], &mut find).is_err());
    }
}
